Iconv tool

NicolasJiaxin opened the pull request:
Fix #146 
This PR adds a tool to directly use the simdutf library from the command line (only transcoding functions for now). In case of an unimplemented transcoding, we use iconv as a fallback. For now, the name of the tool is sutf, but I am open to a better name. The usage is the similar to iconv : `sutf [OPTION...] [−f encoding] [−t encoding] [inputfile ...]` where the only option now is to specify an output file with `[-o output_file]`. If no input is specified, the output is redirected to standard output. There is also `-h` or `--help` and `-l`  or `--list` to display the formats supported by simdutf. There can also be more than one input file.

lemire commented:
Can you run some (silly) benchmarks comparing your tool to iconv? Could be done on AWS.

lemire commented:
We will want to have large file support. So we need to 'eat' the data in blocks... say 4kB (max) at a time. I'd do this after running benchmarks with the current code on both small files, and also files larger than 4kB... 

Ideally, our code should have a finite memory usage, irrespective of the size of the input file.

clausecker commented:
The ideal block size can be determined by calling `stat` on the file descriptor and then checking the `st_blksize` field.

NicolasJiaxin commented:
On individual lipsum files:
**iconv**
```
/home/nicolas/unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       0.9 ms ±   1.4 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     0.8 ms …  77.4 ms    3653 runs
 
/home/nicolas/unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       0.7 ms ±   0.8 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     0.7 ms …  44.4 ms    4173 runs
 
/home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       0.9 ms ±   1.2 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     0.7 ms …  68.0 ms    3983 runs
 
/home/nicolas/unicode_lipsum/lipsum/Hebrew-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Hebrew-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       0.9 ms ±   1.4 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     0.7 ms …  77.4 ms    3936 runs
 
/home/nicolas/unicode_lipsum/lipsum/Hindi-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Hindi-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       0.9 ms ±   2.2 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     0.8 ms … 126.6 ms    3626 runs
 
/home/nicolas/unicode_lipsum/lipsum/Japanese-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Japanese-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       0.7 ms ±   0.1 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     0.7 ms …   7.5 ms    4071 runs
 
/home/nicolas/unicode_lipsum/lipsum/Korean-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Korean-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       0.8 ms ±   1.3 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     0.7 ms …  70.0 ms    4098 runs
 
/home/nicolas/unicode_lipsum/lipsum/Latin-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Latin-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.8 ms ±   2.1 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     0.9 ms … 112.3 ms    3345 runs
 
/home/nicolas/unicode_lipsum/lipsum/Russian-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Russian-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.0 ms ±   1.2 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.0 ms …  66.1 ms    3022 runs
 
```
Command ran: ``for i in \`ls /home/nicolas/unicode_lipsum/lipsum/* | grep utf8\`; do echo $i; hyperfine --warmup 20 "iconv -f UTF-8 -t UTF-16LE $i -o out.txt" -N; done``
**sutf**
```
/home/nicolas/unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.4 ms ±   0.6 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.4 ms …  21.2 ms    2144 runs
 
/home/nicolas/unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.3 ms ±   0.1 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.3 ms …   4.1 ms    2238 runs
 
/home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.3 ms ±   1.1 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.3 ms …  45.1 ms    2343 runs
 
/home/nicolas/unicode_lipsum/lipsum/Hebrew-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Hebrew-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.4 ms ±   0.6 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.3 ms …  25.3 ms    2244 runs
 
/home/nicolas/unicode_lipsum/lipsum/Hindi-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Hindi-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.5 ms ±   0.1 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.4 ms …   4.0 ms    2090 runs
 
/home/nicolas/unicode_lipsum/lipsum/Japanese-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Japanese-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.4 ms ±   1.5 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.3 ms …  70.1 ms    2278 runs
 
/home/nicolas/unicode_lipsum/lipsum/Korean-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Korean-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.4 ms ±   1.4 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.3 ms …  53.1 ms    2287 runs
 
/home/nicolas/unicode_lipsum/lipsum/Latin-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Latin-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.8 ms ±   1.9 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.4 ms …  84.5 ms    2091 runs
 
/home/nicolas/unicode_lipsum/lipsum/Russian-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Russian-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.6 ms ±   0.5 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.5 ms …  22.8 ms    1946 runs
 
```
Command ran: ``for i in `ls /home/nicolas/unicode_lipsum/lipsum/* | grep utf8`; do echo $i; hyperfine --warmup 20 "./build/tools/sutf -f UTF-8 -t UTF-16LE $i -o out.txt" -N; done``

NicolasJiaxin commented:
For three times the same lipsum files:
**iconv**
```
/home/nicolas/unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       3.6 ms ±   4.0 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.5 ms … 121.9 ms    1922 runs
 
/home/nicolas/unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.8 ms ±   0.7 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.2 ms …   9.9 ms    1062 runs
 
/home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.8 ms ±   2.7 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.3 ms …  83.7 ms    986 runs
 
/home/nicolas/unicode_lipsum/lipsum/Hebrew-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Hebrew-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Hebrew-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Hebrew-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.8 ms ±   1.8 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.3 ms …  78.4 ms    2240 runs
 
/home/nicolas/unicode_lipsum/lipsum/Hindi-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Hindi-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Hindi-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Hindi-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.8 ms ±   2.0 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.5 ms …  85.9 ms    1954 runs
 
/home/nicolas/unicode_lipsum/lipsum/Japanese-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Japanese-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Japanese-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Japanese-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.8 ms ±   2.7 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.2 ms … 130.0 ms    2457 runs
 
/home/nicolas/unicode_lipsum/lipsum/Korean-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Korean-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Korean-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Korean-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.8 ms ±   3.1 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.2 ms … 104.5 ms    2186 runs
 
/home/nicolas/unicode_lipsum/lipsum/Latin-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Latin-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Latin-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Latin-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       3.7 ms ±   7.0 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     2.3 ms … 212.5 ms    945 runs
 
/home/nicolas/unicode_lipsum/lipsum/Russian-Lipsum.utf8.txt
Benchmark 1: iconv -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Russian-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Russian-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Russian-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       3.6 ms ±   3.3 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     2.1 ms … 120.8 ms    1370 runs
 
```
Command ran: ``for i in `ls /home/nicolas/unicode_lipsum/lipsum/* | grep utf8`; do echo $i; hyperfine --warmup 20 "iconv -f UTF-8 -t UTF-16LE $i $i $i -o out.txt" -N; done > onefilesutf.txt``
**sutf**
```
/home/nicolas/unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.4 ms ±   0.2 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.4 ms …   8.7 ms    2079 runs
 
/home/nicolas/unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.4 ms ±   0.8 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.3 ms …  33.1 ms    2278 runs
 
/home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Emoji-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.3 ms ±   0.6 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.3 ms …  24.7 ms    2306 runs
 
/home/nicolas/unicode_lipsum/lipsum/Hebrew-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Hebrew-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Hebrew-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Hebrew-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.3 ms ±   0.0 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.3 ms …   1.5 ms    2324 runs
 
/home/nicolas/unicode_lipsum/lipsum/Hindi-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Hindi-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Hindi-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Hindi-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.5 ms ±   1.9 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.4 ms …  82.2 ms    2033 runs
 
/home/nicolas/unicode_lipsum/lipsum/Japanese-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Japanese-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Japanese-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Japanese-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.4 ms ±   1.1 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.3 ms …  49.5 ms    2238 runs
 
/home/nicolas/unicode_lipsum/lipsum/Korean-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Korean-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Korean-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Korean-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.3 ms ±   0.5 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.3 ms …  20.4 ms    2256 runs
 
/home/nicolas/unicode_lipsum/lipsum/Latin-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Latin-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Latin-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Latin-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.9 ms ±   2.5 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.4 ms … 105.7 ms    2062 runs
 
/home/nicolas/unicode_lipsum/lipsum/Russian-Lipsum.utf8.txt
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE /home/nicolas/unicode_lipsum/lipsum/Russian-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Russian-Lipsum.utf8.txt /home/nicolas/unicode_lipsum/lipsum/Russian-Lipsum.utf8.txt -o out.txt
  Time (mean ± σ):       1.6 ms ±   0.8 ms    [User: 0.0 ms, System: 0.0 ms]
  Range (min … max):     1.5 ms …  27.7 ms    1931 runs
 
```
Command ran: ``for i in `ls /home/nicolas/unicode_lipsum/lipsum/* | grep utf8`; do echo $i; hyperfine --warmup 20 "./build/tools/sutf -f UTF-8 -t UTF-16LE $i $i $i -o out.txt" -N; done > onefilesutf.txt``

NicolasJiaxin commented:
On three files, sutf takes about the same time as on one file, so that might mean that a lot of the time is spent on other stuff than actual transcoding.

lemire commented:
> On three files, sutf takes about the same time as on one file, so that might mean that a lot of the time is spent on other stuff than actual transcoding.

Indeed. Try this...

```
perf record "my command"
perf report 
```

By sampling, it should tell you where the time usage goes. Sampling is very crude but it works well enough when there are huge differences such as this one.

By the look of it, we might guess that memory allocation could play a role.

It works also with `iconv`, and I get (for example):

```
Overhead  Command  Shared Object          Symbol
  45.93%  iconv    libc.so.6              [.] __gconv_transform_utf8_internal
  11.74%  iconv    [unknown]              [k] 0xffffba888c97a7e4
   9.65%  iconv    [unknown]              [k] 0xffffba888c6198e0
   8.42%  iconv    [unknown]              [k] 0xffffba888c570ee4
   6.57%  iconv    [unknown]              [k] 0xffffba888c5cb958
   5.94%  iconv    ld-linux-aarch64.so.1  [.] _dl_important_hwcaps
   5.64%  iconv    [unknown]              [k] 0xffffba888c4d45d0
   3.29%  iconv    [unknown]              [k] 0xffffba888c5677bc
   1.22%  iconv    [unknown]              [k] 0xffffba888c551e34
   1.22%  iconv    [unknown]              [k] 0xffffba888c647a30
   0.22%  iconv    [unknown]              [k] 0xffffba888c4d7980
   0.10%  perf-ex  [unknown]              [k] 0xffffba888c4d42c4
   0.06%  perf-ex  [unknown]              [k] 0xffffba888c4d426c
```

NicolasJiaxin commented:
So, it seems indeed that the bottleneck is loading the data. I have played around a bit, but I was not able to bring the time down. Any tips?

lemire commented:
> So, it seems indeed that the bottleneck is loading the data. I have played around a bit, but I was not able to bring the time down. Any tips?

Yes!!! Let me see if I can help.

lemire commented:
I suggest to alias UTF-16 to UTF-16LE.

lemire commented:
There may be a validation issue...

```
./build/tools/sutf -f UTF16LE -t UTF8 unicode_lipsum/lipsum/Russian-Lipsum.utf16.txt
```

Produces `Error with iconv.` after seemingly producing the output.

lemire commented:
It will not help per se, but theoretically,

```C++
std::vector<char16_t> output_buffer(2*size);
```

can be replaced with

```C++
std::unique_ptr<char16_t[]> output_buffer(new char16_t[2*size]);
```

for higher performance...  because the `unique_ptr` way does not require that the input be zeroed.

(It should make no difference here.)

lemire commented:
So if I just have a silly file...

```
$ ls -al hello.txt
-rw-rw-r-- 1 lemire lemire 6 Aug  9 01:27 hello.txt
```

Then iconv is significantly faster...

```
$ hyperfine  'iconv -f UTF-8 -t UTF-16LE hello.txt > /dev/null' --warmup 100
Benchmark 1: iconv -f UTF-8 -t UTF-16LE hello.txt > /dev/null
  Time (mean ± σ):       0.5 ms ±   0.0 ms    [User: 0.5 ms, System: 0.1 ms]
  Range (min … max):     0.5 ms …   0.7 ms    2862 runs
```

compared to...

```
 hyperfine  './build/tools/sutf -f UTF-8 -t UTF-16LE hello.txt > /dev/null' --warmup 100
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE hello.txt > /dev/null
  Time (mean ± σ):       0.9 ms ±   0.0 ms    [User: 0.8 ms, System: 0.1 ms]
  Range (min … max):     0.8 ms …   1.3 ms    2268 runs
```

lemire commented:
Suppose I nuke your main function...

```difff
$ git diff
diff --git a/tools/sutf.cpp b/tools/sutf.cpp
index ee909ef..2860fb6 100644
--- a/tools/sutf.cpp
+++ b/tools/sutf.cpp
@@ -245,12 +245,13 @@ void CommandLine::show_formats() {
 }

 int main(int argc, char* argv[]) {
-  try {
+       return EXIT_SUCCESS;
+  /*try {
     CommandLine cmdline = parse_and_validate_arguments(argc, argv);
     cmdline.run();
     return EXIT_SUCCESS;
   } catch (const std::exception& e) {
       printf("%s\n", e.what());
       return EXIT_FAILURE;
-  }
-}
\ No newline at end of file
+  }*/
+}
```

Then I get...

```
$ hyperfine  './build/tools/sutf -f UTF-8 -t UTF-16LE hello.txt > /dev/null' --warmup 100
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE hello.txt > /dev/null
  Time (mean ± σ):       0.8 ms ±   0.0 ms    [User: 0.8 ms, System: 0.1 ms]
  Range (min … max):     0.8 ms …   1.0 ms    2374 runs
```

... which still appears to be significantly slower than iconv.

NicolasJiaxin commented:
> Produces `Error with iconv.` after seemingly producing the output.

> I suggest to alias UTF-16 to UTF-16LE.

Both issues should be solved with the latest push.

lemire commented:
So I think that the slow starting up is due to dynamic linking... Here is a hack that might drastically speed things up...

```CMake
$ git diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 68d8b1b..f6fc770 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -13,6 +13,7 @@ include(GNUInstallDirs)
 include(cmake/simdutf-flags.cmake)

 set(SIMDUTF_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static-libgcc -static-libstdc++")

 enable_testing()
 add_subdirectory(src)
```

Then I guess, effectively, performance parity...

```
$ hyperfine  './build/tools/sutf -f UTF-8 -t UTF-16LE unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt  > /dev/null' --warmup 100
Benchmark 1: ./build/tools/sutf -f UTF-8 -t UTF-16LE unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt  > /dev/null
  Time (mean ± σ):       0.7 ms ±   0.0 ms    [User: 0.7 ms, System: 0.1 ms]
  Range (min … max):     0.7 ms …   0.9 ms    2415 runs

  Warning: Command took less than 5 ms to complete. Results might be inaccurate.
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

$ hyperfine  'iconv -f UTF-8 -t UTF-16LE unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt  > /dev/null' --warmup 100
Benchmark 1: iconv -f UTF-8 -t UTF-16LE unicode_lipsum/lipsum/Chinese-Lipsum.utf8.txt  > /dev/null
  Time (mean ± σ):       0.7 ms ±   0.0 ms    [User: 0.6 ms, System: 0.1 ms]
  Range (min … max):     0.6 ms …   0.9 ms    2668 runs

  Warning: Command took less than 5 ms to complete. Results might be inaccurate.
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
```

This is before we did any other optimization on the IO.

So clearly, we need to do something with respect to linking.

lemire commented:
(Note that my patch is not recommended, it is just a demonstration.)

lemire commented:
This patch might be reasonable...

```diff
diff --git a/tools/CMakeLists.txt b/tools/CMakeLists.txt
index f51d9ef..c8a59c0 100644
--- a/tools/CMakeLists.txt
+++ b/tools/CMakeLists.txt
@@ -2,6 +2,8 @@ cmake_minimum_required(VERSION 3.15)

 add_executable(sutf sutf.cpp)
 target_link_libraries(sutf PUBLIC simdutf)
-
+if(NOT(MSVC))
+  target_link_options(sutf PRIVATE "-static-libstdc++")
+endif()
```

I am mildly surprised at the overhead of the C++ library.

clausecker commented:
I suspect a number of things.  Probably there is double buffering going on.  Also, as you read the whole file without checking its length first, the code likely employs an exponential backoff strategy, leading to repeated reallocations and hence effectively triple buffering.

How about using straight UNIX file I/O and processing the file in blocks?  There is no need at all to read the whole thing in at once and all the overhead can be avoided.  On the contrary, `iconv` is often used to stream input, so this is generally not even possible.

If desired, I can try to program an example of doing it that way.

lemire commented:
@NicolasJiaxin If you try statistically linking the C++ standard library, it should help quite a bit. Then I recommend trying to use (fixed) stack allocated buffers. There is a bunch of technical optimizations, but our near-term end goal would be to not be embarrassingly slow.

lemire commented:
@clausecker Can you propose a code sample?

lemire commented:
Note that we want to support Windows/Visual Studio.

lemire commented:
@NicolasJiaxin The patch can be improved as follows...

```diff
--- a/tools/CMakeLists.txt
+++ b/tools/CMakeLists.txt
@@ -3,9 +3,9 @@ cmake_minimum_required(VERSION 3.15)
 add_executable(sutf sutf.cpp)
 target_link_libraries(sutf PUBLIC simdutf)

-if(NOT(MSVC))
+if(NOT(MSVC) AND NOT(CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang"))
   target_link_options(sutf PRIVATE "-static-libstdc++")
 endif()
```

So we don't want to use `-static-libstdc++` under macos and visual studio. Possibly, there would be other exclusions.

What we would really want is to check whether the compiler relies on `libstdc++`, but I am not sure how to do that.

lemire commented:
I turned this issue in a blog post...

“Hello world” is slower in C++ than in C (Linux) https://lemire.me/blog/2022/08/09/hello-world-is-slower-in-c-than-in-c-linux/

Because of how our API is designed, we could call the library from C.

lemire commented:
Related https://github.com/simdutf/simdutf/issues/158 and https://github.com/simdutf/simdutf/issues/162

lemire commented:
I have now grown convinced that rewriting the code using the C API (i.e., avoid iostream) will make the problem go away. It is fine to compile using C++ and to use C++, but avoid the C++ headers.

@clausecker : could you provide @NicolasJiaxin with a starting point and how you'd do it?

clausecker commented:
Sure, I'm currently writing on the paper.  Will get to it tomorrow perhaps.

lemire commented:
@clausecker Great.

To clarify, I am not suggesting you do all of the work. Just provide an example of what you'd do so that @NicolasJiaxin has a starting point.

I don't want to statically link the C++ library, by the way. I am convinced that by simply avoiding C++ IO, we can dispense with such nonesense.

Almost all the code written by @NicolasJiaxin is fine. We just need to fix the IO.

(I have written plenty of high performance C++ IO code but it is often much easier to just go with C.)

lemire commented:
I'll see about hacking something so that the C++ code is less terrible. @NicolasJiaxin is relying on input streams which is not the best choice for performance.

lemire commented:
The general lesson is that for performance, never use C++ streams.

lemire commented:
Let me stress that @NicolasJiaxin was using *existing* code in the library for loading the file.

I am off to grab a pizza. I will have a faster version (although not as fast as what @clausecker suggests) soon.

lemire commented:
(@NicolasJiaxin : please verify my result.)

NicolasJiaxin commented:
@lemire Very nice!! I confirm your numbers. You just have a small fix to do because you removed the declaration of `simdutf::to_string(enconding_type bom)`.

clausecker commented:
This looks good!  I believe on Linux at least, you might have now eliminated all double-buffering.  To be really sure, you can try and completely disable buffering with `setbuf(file, nullptr)`.  This is a portable ANSI C call that has also been part of the C++ standard since the beginning (`std::setbuf`), so it should work just fine on Windows.

With buffering disabled, the stdio code will read directly into your buffer.

I also recommend taking out the code that checks for whether the file exists.  The `fopen` call will fail anyway if the file does not, so it's redundant and may cause a confusing mismatch in error reporting depending on the error condition.  Also consider using `perror` to report IO errors as that call automatically prints the message corresponding to the current `errno` value from the failed `stdio` call, leading to better error reporting.

As for `fseek` and `ftell`, consider using `fseeko` and `ftello` for better large file support (these use `off_t` instead of `long`). You can pretty much do

    #ifdef __WIN32
    #define fseeko _fseeki64
    #define ftello _ftelli64
    #endif

to address the difference between Win32 and UNIX.

lemire commented:
@NicolasJiaxin I have also added another tweak for cleaner CMake. I am now convinced that C++ dynamic linking may add up to 1 ms of overhead (despite hundreds of people telling me otherwise) as an overhead (each time you start the process). Going static drastically reduce this code.

I think that @clausecker has good ideas.

For large files, I would not do things the way we do it now. You would not want to allocate gigabytes of memory... :-)

Here what I suggest you do Nicolas: try benchmarking on our usual files, against iconv, after applying something like my patch. Hopefully, we are beating at at least matching `iconv` throughout with a statically linked binary.

Next, of course, it will be interesting to see us beat iconv even more when adopting @clausecker 's code (AVX-512).

I think that this brings up issue: https://github.com/simdutf/simdutf/issues/158

If we could drastically reduce our dependence on the C++ library... then we would not need to link statistically against it.  (To be clear, we would still use C++, just not the standard library.)

lemire commented on tools/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+add_executable(sutf sutf.cpp)
+target_link_libraries(sutf PUBLIC simdutf)
+
+if(NOT(MSVC) AND NOT(CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang"))
```
In my PR to your PR, I have a better proposal for this condition.

lemire commented on tools/sutf.cpp:
```diff
@@ -0,0 +1,256 @@
+  }
+}
+
+void CommandLine::iconv_fallback() {
```
The iconv_fallback feels like a hack, though it is a good first start.

lemire commented:
The iconv link is too weak. We want to do better (e.g., https://cmake.org/cmake/help/latest/module/FindIconv.html).

lemire commented:
I am working on a more solid approach regarding iconv... give me a few minutes...

lemire commented:
Have a look at https://github.com/simdutf/simdutf/pull/164

lemire commented:
Let me have a look at that error...

lemire commented:
@NicolasJiaxin Try removing or commenting out the `SIMDUTF_POP_DISABLE_WARNINGS` lines. For some odd reason, they create trouble. I am not sure why. But removing them should be safe if the result is a clean build.

NicolasJiaxin commented:
Comparing iconv and sutf on single lipsum files.
Sample command for sutf: hyperfine "./build/tools/sutf -f UTF-32LE -t UTF-16LE lipsum_file -o /dev/null" --warmup 100 -N
Sample command for iconv: hyperfine "iconv -f UTF-32LE -t UTF-16LE lipsum_file -o /dev/null" --warmup 100 -N
||Average sutf (ms)|Average iconv (ms)|
|-|-|-|
|**UTF-8 to UTF-16LE**| 0.5 | 0.689 |
|**UTF-8 to UTF-32**| 0.511 | 0.744 |
|**UTF-16LE to UTF-8**| 0.5 | 0.7 |
|**UTF-16LE to UTF-16BE**| 0.5 | 0.633 |
|**UTF-16LE to UTF-32**| 0.522 | 0.678 |
|**UTF-32 to UTF-8**| 0.522 | 0.656 |
|**UTF-32 to UTF-16LE**| 0.522 | 0.644 |

Lipsum files are pretty big, from ~45KB to ~226KB. I can rerun on smaller files and also check with perf where the time usage is spent.

NicolasJiaxin commented:
From a quick test on a small file ~1KB, iconv and sutf were equal, taking 0.4ms according to hyperfine.

NicolasJiaxin commented:
I am not sure what perf is indicating, but it does not look like most of the time usage is spent on transcoding. But now, sutf is faster than iconv at least.

lemire commented:
> But now, sutf is faster than iconv at least.

It would be interesting to see how this varies based of the file size.

NicolasJiaxin commented:
I generated random UTF-8 files of different sizes and transcoded to UTF-16LE with sutf and iconv. The results:
|File size|sutf (ms)|iconv (ms)|
|-|-|-|
|**16B**|0.4|0.4|
|**128B**|0.4|0.4|
|**512B**|0.4|0.4|
|**1KB**|0.4|0.4|
|**4KB**|0.4|0.4|
|**16KB**|0.5|0.5|
|**100KB**|0.7|1.2|
|**500KB**|1.8|4.4|
|**2MB**|6.3|16.8|

For small files (at most less than 100KB), sutf and iconv perform similarly. And beyond that, sutf is faster although this may not be relevant later when we will process the files in small blocks. From these numbers, it looks like about 4KB is the optimal block size (at least for this random UTF-8 to UTF-16LE).

lemire commented on tools/sutf.cpp:
```diff
@@ -0,0 +1,313 @@
+  if (output_file.empty()) {
+    run_procedure(stdout);
+  } else {
+    SIMDUTF_DISABLE_DEPRECATED_WARNING 
```
@NicolasJiaxin 

I have figured it out.
You have to do...
```
  SIMDUTF_PUSH_DISABLE_WARNINGS
  SIMDUTF_DISABLE_DEPRECATED_WARNING
 something here
  SIMDUTF_POP_DISABLE_WARNINGS
```

Could you fix these? I think it is my fault, I called `SIMDUTF_DISABLE_DEPRECATED_WARNING` without first calling `SIMDUTF_PUSH_DISABLE_WARNINGS` which caused Visual Studio to get angry.

lemire commented:
> For small files (at most less than 100KB), sutf and iconv perform similarly. And beyond that, sutf is faster although this may not be relevant later when we will process the files in small blocks. From these numbers, it looks like about 4KB is the optimal block size (at least for this random UTF-8 to UTF-16LE).

I like these numbers !!!

Can you try the small block approach? At least a first stab? You could do it as a separate branch and compare. You'd hope that such code would be faster than even sutf for large files.

NicolasJiaxin commented:
Yes I can try!

lemire commented:
You could also try to fit your numbers with a linear function a x + b. It should be a good fit for the timings. (Simple linear regression should do!)

NicolasJiaxin commented:
I plotted the following data:
|File size|sutf (ms)|iconv (ms)|
|-|-|-|
|16KB|0.4||
|16KB|0.4||
|16KB|0.4||
|16KB|0.4||
|16KB|0.4||
|16KB|0.4||
![image](https://user-images.githubusercontent.com/83141563/185209521-4f875d7f-0d99-4b44-aea2-66e0d65529cd.png)

lemire commented:
@NicolasJiaxin  Look, mathematics work in the real world !!!

lemire commented:
So in the limit, sutf (in this particular test) is 2.7x faster than iconv.

NicolasJiaxin commented:
> So in the limit, sutf (in this particular test) is 2.7x faster than iconv.

I will rerun this test when I will implement the small block size approach.

lemire commented:
> As the conversion is only a part of the total runtime, the actual performance improvement over an iconv() call is likely a lot more than that.

Though it is not entirely complete, we do have iconv (the library) as part of our standard benchmarks now!!!

lemire commented:
@clausecker 

I would also expect that iconv/sutf would undersell our speedups, but maybe not as much as I expect since 2.7 is already a good gain (prior to your recent work). Here is a sample result...

```
$ ./build/benchmarks/benchmark -P convert_utf8_to_utf16+haswell -P convert_utf8_to_utf16+iconv -F unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt
We define the number of bytes to be the number of *input* bytes.
We define a 'char' to be a code point (between 1 and 4 bytes).
===========================
Using ICU version 70.1
testcases: 1
input detected as UTF8
current system detected as haswell
===========================
convert_utf8_to_utf16+haswell, input size: 81685, iterations: 2000, dataset: unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt
   2.362 GB/s (1.1 %)    1.323 Gc/s     1.78 byte/char
convert_utf8_to_utf16+iconv, input size: 81685, iterations: 2000, dataset: unicode_lipsum/lipsum/Arabic-Lipsum.utf8.txt
   0.702 GB/s (5.8 %)    0.393 Gc/s     1.78 byte/char
```

So more than 3x on that particular system that I used.


Details matter, of course.  We will need a full investigation.

lemire commented:
Thanks for the update.

lemire commented:
@NicolasJiaxin I am mildly concerned with the fact that you are seemingly allocating a new output buffer repeatedly...

```
std::unique_ptr<char16_t[]> output_buffer(new char16_t[size]);
```

You could have just one of those and reuse it. Just budget it so that it is large enough. 

What you are doing is probably efficient... the memory allocator is not going to be an important burden, and may be effectively free (always reusing the same memory), however it is *inelegant*.

lemire commented:
It would be nice to have benchmarks (over files of various sizes).

NicolasJiaxin commented:
I changed the code to use just one output buffer that is always large enough. I also corrected a few bugs and tested more thoroughly sutf. I was able to transcode all lipsum files exactly (from UTF-8 to UTF-16LE and UTF-32, from UTF-16LE to UTF-8 and UTF-32, from UTF-32 to UTF-8 to UTF-16LE) with the minor caveat that our functions are not BOM-aware so sometimes the output file has an extra BOM or is missing a BOM.

NicolasJiaxin commented:
@lemire Do you have an idea for what the value of CHUNK_SIZE should be?

NicolasJiaxin commented:
From trying a few  values for CHUNK_SIZE, I think 64KB is the best. I compared 8 different chunk sizes (1KB, 4KB, 16KB, 32KB, 64KB, 128KB, 512KB and 1MB) and for each, I used hyperfine to benchmark on 11 UTF-8 randomly generated files of different sizes (about 16KB, 32KB, 64KB, 100KB, 250KB, 500KB, 1MB, 2MB, 4MB, 10MB and 20MB, the files were the same for all chunk sizes). Performing linear regression we get (time vs. file size),

|CHUNK_SIZE|slope|
|1KB|2.752|
|4KB|2.616|
|16KB||
|16KB||
|16KB||

NicolasJiaxin commented:
64KB performs better on smaller files, 16KB to 1MB, (10.1ms faster) than 128KB. This represents ~10-15% decrease in time and both 64KB and 128KB perform very similarly otherwise. So I would say 64KB is best (at least according to my test for random UTF-8 and UTF-8 to UTF-16LE transcoding). Note that this is much better than before adding file streaming (linear regression yielded y = 2.959x + 343.9). The difference is noticeable in the data for file of 1MB and more and there is almost no difference in small files.

NicolasJiaxin commented:
I will merge `toolstream` into this PR.

lemire commented:
Brilliant. You did exactly what I would have done regarding the buffer size. It is likely to be system-specific and it may not be possible to come up with an ideal number for all systems, but 64kB sounds good. We could also add a flag later to allow the user to change this default.

I will investigate the visual studio issue at once.

NicolasJiaxin commented:
> I will investigate the visual studio issue at once.

I think it is just a format specifier issue in a few printf statements. I will fix it.

lemire commented:
>  format string '%lu' requires an argument of type 'unsigned long', but variadic argument 1 has type 'size_t'

Try %zu

lemire commented:
https://www.tutorialspoint.com/what-is-the-correct-way-to-use-printf-to-print-a-size-t-in-c-cplusplus

lemire commented:
It passes all the tests !!!

NicolasJiaxin commented:
Yes!! Thank you!

lemire commented:
Merging.
